### PR TITLE
Provide a gRPC alternative of HTTP's /version API

### LINKED ISSUE #14599: Provide a gRPC alternative of HTTP's `/version` API
### What would you like to be added?

There is a version API in HTTP part which returns the `server` / `cluster` / `storage` versions:
https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/etcdhttp/version.go

It seems there is no corresponding API in the gRPC part.
The closest one I found is the `Maintenance.Status` API:
https://github.com/etcd-io/etcd/blob/e24402d39fd08ebca46beeb1f16d214e37fea74f/api/etcdserverpb/rpc.proto#L1140
https://github.com/etcd-io/etcd/blob/e24402d39fd08ebca46beeb1f16d214e37fea74f/server/etcdserver/api/v3rpc/maintenance.go#L228

But it doesn't have the cluster version field.

### Why is this needed?

By providing the gRPC alternative of HTTP's `/version` API, we can access all the features of etcd via gRPC connection, without using an extra HTTP client.

### PR DESCRIPTION
**Pull request by spacewander**
Fix #14599

Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### CODE REVIEW

**Comment by spacewander** [comment]
@ahrtr 
I am not familiar with etcd, could you tell me where should I put the test in?

File: api/etcdserverpb/rpc.proto [hunk 1]
```diff
@@ -1161,6 +1161,8 @@ message StatusResponse {
   bool isLearner = 10 [(versionpb.etcd_version_field)="3.4"];
   // storageVersion is the version of the db file. It might be get updated with delay in relationship to the target cluster version.
   string storageVersion = 11 [(versionpb.etcd_version_field)="3.6"];
+  // clusterVersion is the minimum cluster protocol version used in the cluster.
+  string clusterVersion = 12 [(versionpb.etcd_version_field)="3.7"];
```
Comment (ahrtr) [inline api/etcdserverpb/rpc.proto:1165]:
```suggestion
  string clusterVersion = 12 [(versionpb.etcd_version_field)="3.6"];
```

**Comment by ahrtr** [comment]
> @ahrtr I am not familiar with etcd, could you tell me where should I put the test in?

We will add mix version (e.g. 3.6 mix with 3.5) testing recently. We will take care of the test for this PR as well.

**Comment by spacewander** [comment]
@ahrtr 
Updated

**Comment by ptabor** [comment]
> > @ahrtr I am not familiar with etcd, could you tell me where should I put the test in?
> 
> We will add mix version (e.g. 3.6 mix with 3.5) testing recently. We will take care of the test for this PR as well.

I would recommend adding just regression test in:  `tests/common/status_test.go` that checks that the returned version is consistent with `/api/version/version.go`

**Comment by spacewander** [comment]
The `/api/version/version.go` doesn't contain cluster/storage version. Do you mean adding an e2e test to make sure we can get all fields in HTTP `/version` by using the gRPC call `cc.Status()`?

**Comment by ahrtr** [comment]
To clarify @ptabor 's comment above https://github.com/etcd-io/etcd/pull/14614#issuecomment-1289006142, please add a common test case in [status_test.go](https://github.com/etcd-io/etcd/blob/main/tests/common/status_test.go), to check that the `version` matches `clusterVersion` returned by the endpoint `status` API.

**Comment by spacewander** [comment]
I check the error log. 

```
2022-10-26T02:27:48.7827993Z         logger.go:130: 2022-10-26T02:25:19.753Z    INFO    started server. {"name": "TestStatusPeerTLS-test-1", "pid": 27529}
2022-10-26T02:27:48.7828561Z         logger.go:130: 2022-10-26T02:25:19.753Z    INFO    started server. {"name": "TestStatusPeerTLS-test-2", "pid": 27528}
2022-10-26T02:27:48.7829120Z         logger.go:130: 2022-10-26T02:25:19.763Z    INFO    started server. {"name": "TestStatusPeerTLS-test-0", "pid": 27537}
2022-10-26T02:27:48.7829485Z         status_test.go:79: mismatch cluster versions. expected:, got:3.6.0
```

When `TestStatus/PeerTLS` is run, the `ms.vs.GetClusterVersion()` seems return nil value, which causes the  ClusterVersion in the gRPC response to be empty.

However, when I run this locally four times, this test always passes:
```
 ￥ go test --tags e2e go.etcd.io/etcd/tests/v3/common -run ^TestStatus$
ok      go.etcd.io/etcd/tests/v3/common 12.806s
 tests (azq2) 0
 ￥ go test --tags e2e -count 1 go.etcd.io/etcd/tests/v3/common -run ^TestStatus$
ok      go.etcd.io/etcd/tests/v3/common 11.912s
 tests (azq2) 0
 ￥ go test --tags e2e -count 1 go.etcd.io/etcd/tests/v3/common -run ^TestStatus$
ok      go.etcd.io/etcd/tests/v3/common 11.698s
 tests (azq2) 0
 ￥ go test --tags e2e -count 1 go.etcd.io/etcd/tests/v3/common -run ^TestStatus$
ok      go.etcd.io/etcd/tests/v3/common 10.232s
```

File: tests/common/status_test.go [hunk 2]
```diff
@@ -53,6 +56,29 @@ func TestStatus(t *testing.T) {
...
+					} else {
+						if version.LessThan(*ver, *minVer) {
+							minVer = ver
+						}
+					}
+				}
+				for _, r := range rs {
+					if r.ClusterVersion != minVer.String() {
+						t.Fatalf("mismatch cluster versions. expected:%s, got:%s", r.ClusterVersion, minVer.String())
+					}
```
Comment (ahrtr) [inline tests/common/status_test.go:80]:
```suggestion
					assert.Equal(t, minVer.String(), r.ClusterVersion, "unexpected cluster version")
```

File: tests/common/status_test.go [hunk 3]
```diff
@@ -53,6 +56,29 @@ func TestStatus(t *testing.T) {
 				if len(rs) != len(memberIds) {
 					t.Fatalf("found duplicated members")
 				}
+
+				var minVer *semver.Version
+				for _, r := range rs {
+					if r.Version != version.Version {
+						t.Fatalf("mismatch versions. expected:%s, got:%s", version.Version, r.Version)
+					}
```
Comment (ahrtr) [inline tests/common/status_test.go:64]:
```suggestion
					assert.Equal(t, version.Version , r.Version, "unexpected server version")
					
```

File: tests/common/status_test.go [hunk 4]
```diff
@@ -19,6 +19,9 @@ import (
 	"testing"
 	"time"
 
+	"github.com/coreos/go-semver/semver"
+
```
Comment (ahrtr) [inline tests/common/status_test.go:23]: add
"github.com/stretchr/testify/assert"

**Comment by ahrtr** [comment]
You also need to update the `etcdctl` to display the new field "`clusterVersion`".

**Comment by spacewander** [comment]
CI failed when running `TestStatus/PeerAutoTLS`:
```
2022-10-28T07:11:35.3023127Z         status_test.go:77:
2022-10-28T07:11:35.3023518Z                    Error Trace:    status_test.go:77
2022-10-28T07:11:35.3024036Z                                                            execute.go:38
2022-10-28T07:11:35.3024466Z                                                            asm_386.s:1326
2022-10-28T07:11:35.3024776Z                    Error:          Not equal:
2022-10-28T07:11:35.3025200Z                                    expected: "3.6.0"
2022-10-28T07:11:35.3025584Z                                    actual  : ""
2022-10-28T07:11:35.3025868Z
2022-10-28T07:11:35.3026194Z                                    Diff:
2022-10-28T07:11:35.3026626Z                                    --- Expected
2022-10-28T07:11:35.3027000Z                                    +++ Actual
2022-10-28T07:11:35.3027411Z                                    @@ -1 +1 @@
2022-10-28T07:11:35.3027794Z                                    -3.6.0
2022-10-28T07:11:35.3028094Z                                    +
2022-10-28T07:11:35.3028460Z                    Test:           TestStatus/PeerAutoTLS
2022-10-28T07:11:35.3028880Z                    Messages:       unexpected cluster version
2022-10-28T07:11:35.3029466Z         logger.go:130: 2022-10-28T07:09:21.475Z    INFO    closing test cluster...
```

However, running the test on my machine can always pass:
```
go test --tags e2e -count 1 go.etcd.io/etcd/tests/v3/common -run ^TestStatus$
ok      go.etcd.io/etcd/tests/v3/common 10.519s
```

It is pretty strange.

**Pull request closed without merging**